### PR TITLE
Rework network module

### LINKED ISSUE #309: Internet page improvements
- [ ] Show the sum of other MAC addresses when displaying MAC address requests.
- [x] Show wifi username on internet access table

### LINKED ISSUE #370: Remove confusing characters from password generation
It has happened several times that users are having troubles connecting to the network. Usually it is only an issue of not typing the password correctly. Like confusing `y` with `z` (common English vs common Hungarian keyboard), or confusing `I` with `l` or `1`. We should probably look into the types of characters that can cause confusion and remove them from WiFi password generation.

I think it could drastically improve user experience without the users even noticing it.

### PR DESCRIPTION
**Pull request by kdmnk**
- Split internetcontroller into InternetController, AdminInternetController and MacAddressController.
- Deleted internet-user role
- Deleted auto-approved MACs for https://github.com/EotvosCollegium/mars/security/advisories/GHSA-x3j4-cf37-2fpr
- Tried to use REST structures
- Test for all the controller actions
- Improved tables on the pages
- Added wifi username and password to admin page
- #370 
- And some other changes that should have been in a separate PR, sorry: generated ide helpers, commented out triggers in migrations, and formatting
- New internet fault report form
- Improved email testing across test cases

### CODE REVIEW

File: app/Policies/MacAddressPolicy.php [hunk 1]
```diff
@@ -16,57 +15,31 @@ public function before(User $user)
...
+     * Determine whether the user can create mac address instances.
      *
      * @param User $user
-     * @return mixed
+     * @return bool
      */
     public function create(User $user): bool
     {
-        return true;
+        return $user->can('handle', $user->internetAccess);
```
Comment (viktorcsimma) [inline app/Policies/MacAddressPolicy.php]: Does this mean that the user has an internet access? It is a bit confusing for me.

Comment (viktorcsimma) [reply app/Policies/MacAddressPolicy.php]: Maybe it would be better to sign it like, "if the user's internet access does exist and has not expired".

File: app/Models/Internet/InternetAccess.php [hunk 2]
```diff
@@ -67,6 +120,34 @@ public function setWifiCredentials($username = null): string
...
+
+    /**
+     * Generate a "random" alphanumeric string, with the confusing characters filtered out.
+     * This is a modified version of Str::random() from Laravel.
+     *
+     * @return string
+     */
+    private static function generateWifiPassword(): string
+    {
+        $disabled_chars = ['/', '+', '=', 'o', '0', 'z', 'y', 'l', '1', 'i', 'I'];
```
Comment (viktorcsimma) [inline app/Models/Internet/InternetAccess.php]: Here, capital 'Z' and 'Y' are missing; I was able to reproduce them in generated passwords, too.

File: app/Models/Internet/InternetAccess.php [hunk 3]
```diff
@@ -67,6 +120,34 @@ public function setWifiCredentials($username = null): string
...
+
+        $string = '';
+
+        while (($len = strlen($string)) < self::PASSWORD_LENGTH) {
+            $size = self::PASSWORD_LENGTH - $len;
+
+            $bytes = random_bytes($size);
+
+            $string .= substr(str_replace($disabled_chars, '', base64_encode($bytes)), 0, $size);
```
Comment (viktorcsimma) [inline app/Models/Internet/InternetAccess.php:167]: What does the `substr` call do here? If I understand right, this already is at most `$size` long.

File: app/Models/EpistolaNews.php [hunk 4]
```diff
@@ -6,6 +6,56 @@
...
+ *
+ * @property int $id
+ * @property int|null $uploader_id
+ * @property string $title
+ * @property string|null $subtitle
+ * @property string|null $description
+ * @property \Illuminate\Support\Carbon|null $date
+ * @property \Illuminate\Support\Carbon|null $time
+ * @property \Illuminate\Support\Carbon|null $end_date
+ * @property string|null $picture_path
```
Comment (viktorcsimma) [inline app/Models/EpistolaNews.php:20]: Just a question for myself; what are these?

Comment (viktorcsimma) [inline app/Models/EpistolaNews.php:20]: Just a question for myself; what are these?

File: database/migrations/2019_10_19_153240_create_trigger.php [hunk 5]
```diff
@@ -9,36 +9,36 @@ class CreateTrigger extends Migration
...
                ON users.id = print_accounts.user_id
             WHERE
                print_accounts.user_id IS NULL;');
-        DB::unprepared('
-        CREATE TRIGGER trigger_create_print_account_for_user
-            AFTER INSERT ON users 
-            FOR EACH ROW
-            INSERT INTO print_accounts(user_id) VALUES (NEW.id);
-        ');
+        //        DB::unprepared('
```
Comment (viktorcsimma) [inline database/migrations/2019_10_19_153240_create_trigger.php:20]: Maybe we should delete these comments altogether; we can still restore them from git if we need them.

Comment (viktorcsimma) [reply database/migrations/2019_10_19_153240_create_trigger.php:20]: Same at the other places.

File: database/migrations/2023_01_09_165653_drop_triggers.php [hunk 6]
```diff
@@ -13,11 +13,11 @@
...
-        DB::unprepared('DROP TRIGGER trigger_insert_print_account_history_free_pages');
-        DB::unprepared('DROP TRIGGER trigger_update_print_account_history_free_pages');
+        //        DB::unprepared('DROP TRIGGER trigger_create_print_account_for_user');
+        //        DB::unprepared('DROP TRIGGER trigger_create_internet_access_for_user');
+        //        DB::unprepared('DROP TRIGGER trigger_print_account_history_balance');
+        //        DB::unprepared('DROP TRIGGER trigger_insert_print_account_history_free_pages');
+        //        DB::unprepared('DROP TRIGGER trigger_update_print_account_history_free_pages');
     }
 
     /**
```
Comment (viktorcsimma) [inline database/migrations/2023_01_09_165653_drop_triggers.php:23]: Or is there any argument against simply deleting the entire migration file? If yes, we could add a different one reversing it (so readding the triggers).

File: resources/lang/hu/internet.php [hunk 7]
```diff
@@ -19,13 +18,14 @@
...
+    'registered_devices_info' => 'A vezetékes internet használatához előbb regisztrálnod kell az eszközödet. Ezt biztonsági okokból manuálisan el kell fogadniuk a rendszergazdáknak, így szíves türelmedet kérjük. Igyekszünk minél hamarabb elfogadni a kérelmeket.',
     'rejected' => 'Elutasított',
     'report' => 'A probléma leírása',
     'report_fault' => 'Hiba bejelentése',
-    'report_fault_desc' => 'Ha bármi probléma merül fel az interneteléréseddel kapcsolatban, kérlek jelentsd be az alábbi űrlap kitöltésével. Próbáld meg a hibát minél pontosabban leírni és jelezd, mikor vagy a szobában, hogy a rendszergazdák fel tudjanak keresni.',
+    'report_fault_desc' => 'Ha bármi probléma merül fel az interneteléréseddel kapcsolatban, kérlek jelentsd be az alábbi űrlap kitöltésével. Próbáld meg a hibát minél pontosabban leírni (mióta, hogyan jelentkezik a probléma, van-e hibaüzenet, próbáltad-e más eszközön, stb.) és jelezd, mikor vagy a szobában, hogy a rendszergazdák fel tudjanak keresni.',
```
Comment (viktorcsimma) [inline resources/lang/hu/internet.php]: We could also add something like "what you have already tried to solve the issue". This would save us from many unnecessary rounds.

File: resources/lang/en/internet.php [hunk 8]
```diff
@@ -19,9 +18,10 @@
     'internet_access' => 'Internet access',
     'laptop' => 'Laptop',
     'mac_address' => 'MAC address',
+    'mac_comment' => 'eg. Asus Laptop, Ethernet converter',
     'no_internet' => 'You are not allowed to access the internet at the moment.',
     'nothing_to_show' => 'Nothing to show',
-    'registered_devices_info' => 'You must register your device to use wired connection. You can approve a certain number of devices automatically. If you need more, then the administrators must approve it after your registration; you can provide a reason in the comment field.',
+    'registered_devices_info' => 'To use wired internet, you must first register your device. For security reasons, system administrators have to accept this manually, so we kindly asks for your patience. We try to accept requests as soon as possible.',
```
Comment (viktorcsimma) [inline resources/lang/en/internet.php]: Just a grammatical error: `kindly ask`.

File: resources/lang/hu/internet.php (hunk 7 above)
Comment (viktorcsimma) [reply resources/lang/hu/internet.php]: Something else: should we also extend this text in the English version, too?

File: resources/views/network/admin/mac_addresses.blade.php [hunk 9]
```diff
@@ -0,0 +1,150 @@
+<span
+    class="card-title">Regisztrált vezetékes eszközök kezelése
+    @notification(\App\Models\Internet\MacAddress::class)
+</span>
+<blockquote class="error">
+    A MAC címek jóváhagyása előtt ellenőrizd, hány címe van már a felhasználónak, és hogy nem routerről van-e szó
+    (<a href="https://mac-address.alldatafeeds.com/mac-address-lookup" target="_blank">ellenőrizd pl. itt</a>).
+    Módosítsd a megjegyzést, ha szükséges.
```
Comment (viktorcsimma) [inline resources/views/network/admin/mac_addresses.blade.php]: We could include some kind of warning about what @horcsinbalint has said about `ff:ff:ff:ff:ff` and friends.

File: app/Models/Role.php [hunk 10]
```diff
@@ -293,7 +306,7 @@ public function color(): string
             self::LOCALE_ADMIN => 'amber',
```
Comment (viktorcsimma) [inline app/Models/Role.php:305]: Here, `self::INTERNET_USER` is still there; and it complains about that when I try to access the users index.

**Review by viktorcsimma — CHANGES_REQUESTED** [review]
It seems to be nice:) One important thing besides my comments: I get a 403 when I try to access the routers' index. Could you investigate that?

Btw, thanks for trusting me with the request:) sorry if I ask something that I should know.

File: app/Policies/MacAddressPolicy.php (hunk 1 above)
Comment (kdmnk) [reply app/Policies/MacAddressPolicy.php]: This only redirects it to the InternetAccessPolicy's handle which only checks currently that it can not create a Mac for other users. Which is true in every case in this call :D I'll just delete it. (The internet access should exist by default, but if not it also returns 403 for null. The expired check is not that necessary, just creates more tension.)

File: app/Models/Internet/InternetAccess.php (hunk 3 above)
Comment (kdmnk) [reply app/Models/Internet/InternetAccess.php:167]: idk, as I said in the comments, I just modified laravel's algorithm, which I trust :D

File: app/Models/EpistolaNews.php (hunk 4 above)
Comment (kdmnk) [reply app/Models/EpistolaNews.php:20]: These should help your IDE find the Laravel-generated properties and their types. These are generated with PHP artisan ide-helper (an external tool)

File: database/migrations/2019_10_19_153240_create_trigger.php (hunk 5 above)
Comment (kdmnk) [reply database/migrations/2019_10_19_153240_create_trigger.php:20]: Our git history is not the nicest the search in... They won't harm us here ;) And we can squash the migrations into one file now containing a final db structure, I just could not make it work yet.

File: app/Http/Controllers/Network/AdminInternetController.php [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+     * @return LengthAwarePaginator
+     * @throws AuthorizationException
+     */
+    public function indexInternetAccesses(): LengthAwarePaginator
+    {
+        $this->authorize('handleAny', InternetAccess::class);
+
+        return TabulatorPaginator::from(
+            InternetAccess::query()
+                ->join('users as user', 'user.id', '=', 'user_id')
```
Comment (sszajbely) [inline app/Http/Controllers/Network/AdminInternetController.php:45]: Is this `join` needed when using `with` below?

File: app/Http/Controllers/Network/AdminInternetController.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+        ]);
+
+        if ($request->get('has_internet_until') != null) {
+            $newDate = $request->get('has_internet_until');
+        } else {
+            $newDate = InternetAccess::getInternetDeadline();
+        }
+        $internetAccess->update(['has_internet_until' => $newDate]);
+        return Carbon::parse($newDate);
+    }
```
Comment (sszajbely) [inline app/Http/Controllers/Network/AdminInternetController.php]: Maybe this instead?
```suggestion
        if ($request->get('has_internet_until') != null) {
            $newDate = Carbon::parse($request->get('has_internet_until'));
        } else {
            $newDate = InternetAccess::getInternetDeadline();
        }
        $internetAccess->update(['has_internet_until' => $newDate]);
        return $newDate;
    }
```

File: app/Http/Controllers/Network/InternetController.php [hunk 13]
```diff
@@ -4,250 +4,68 @@
...
+        $request->validate([
             'report' => 'required|string',
             'user_os' => 'required|string',
         ]);
-        $validator->validate();
 
-        foreach (User::withRole(Role::SYS_ADMIN)->get() as $admin) {
-            Mail::to($admin)->queue(new InternetFault($admin->name, user()->name, $request->report, $request->user_os));
+        foreach (User::admins() as $admin) {
+            Mail::to($admin)->queue(new InternetFault($admin->name, user()->name, $request->get('report'), $request->get('user_os')));
```
Comment (sszajbely) [inline app/Http/Controllers/Network/InternetController.php]: Question for all of these kind of emails. Do we really need the emails stating the name of the admin? Couldn't it just say 'Dear Admin, ...'? If that was okay we could use `Mail::to(User::admins())->...` instead.

File: app/Http/Controllers/Network/MacAddressController.php [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+     * @return LengthAwarePaginator
+     * @throws AuthorizationException
+     */
+    public function index(Request $request): LengthAwarePaginator
+    {
+        $this->authorize('handleAny', InternetAccess::class);
+
+        return TabulatorPaginator::from(
+            MacAddress::query()
+                ->join('users as user', 'user.id', '=', 'user_id')
```
Comment (sszajbely) [inline app/Http/Controllers/Network/MacAddressController.php:35]: Same as above

File: app/Http/Controllers/Network/MacAddressController.php [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+     * @throws ValidationException
+     */
+    public function store(Request $request): RedirectResponse
+    {
+        $this->authorize('create', MacAddress::class);
+
+        Validator::validate($request->all(), [
+            'comment' => 'required|max:255',
+            'mac_address' => ['required', 'regex:/((([a-fA-F0-9]{2}[-:]){5}([a-fA-F0-9]{2}))|(([a-fA-F0-9]{2}:){5}([a-fA-F0-9]{2})))/i'],
+        ]);
```
Comment (sszajbely) [inline app/Http/Controllers/Network/MacAddressController.php]:
```suggestion
        $request->validate([
            'comment' => 'required|max:255',
            'mac_address' => ['required', 'regex:/((([a-fA-F0-9]{2}[-:]){5}([a-fA-F0-9]{2}))|(([a-fA-F0-9]{2}:){5}([a-fA-F0-9]{2})))/i'],
        ]);
```

File: app/Http/Controllers/Network/MacAddressController.php [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+        MacAddress::create([
+            'user_id' => user()->id,
+            'comment' => $request->input('comment'),
+            'mac_address' => $request->input('mac_address')
+        ]);
+
+        foreach (User::admins() as $admin) {
+            Mail::to($admin)->send(new MacNeedsApproval($admin->name, user()->name));
+        }
```
Comment (sszajbely) [inline app/Http/Controllers/Network/MacAddressController.php]: Same as above

File: app/Policies/MacAddressPolicy.php [hunk 17]
```diff
@@ -16,57 +15,31 @@ public function before(User $user)
...
     /**
-     * Determine whether the user can update the mac address.
+     * Determine whether the user can approve a mac address.
      *
      * @param User $user
-     * @param MacAddress $macAddress
      * @return mixed
      */
-    public function update(User $user, MacAddress $macAddress): bool
+    public function updateState(User $user): bool
```
Comment (sszajbely) [inline app/Policies/MacAddressPolicy.php]: Why did you rename this?

File: app/Http/Controllers/Network/MacAddressController.php [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        $this->authorize('updateState', $macAddress);
+
+        $request->validate([
+            'comment' => 'nullable|string|max:255',
+            'state' => 'nullable|in:' . implode(',', MacAddress::STATES)
+        ]);
+
+        $macAddress->update($request->only(['state', 'comment']));
+
+        return $macAddress->refresh();
```
Comment (sszajbely) [inline app/Http/Controllers/Network/MacAddressController.php]: I don't think the refresh is needed (maybe I'm wrong)

File: app/Http/Controllers/Network/MacAddressController.php [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    public function update(Request $request, MacAddress $macAddress): MacAddress
+    {
+        $this->authorize('updateState', $macAddress);
+
+        $request->validate([
+            'comment' => 'nullable|string|max:255',
+            'state' => 'nullable|in:' . implode(',', MacAddress::STATES)
+        ]);
+
+        $macAddress->update($request->only(['state', 'comment']));
```
Comment (sszajbely) [inline app/Http/Controllers/Network/MacAddressController.php]:
```suggestion
        $data = $request->validate([
            'comment' => 'nullable|string|max:255',
            'state' => 'nullable|in:' . implode(',', MacAddress::STATES)
        ]);

        $macAddress->update($data);
```
It's a bit less code, but maybe less readable.

File: app/Models/Internet/InternetAccess.php [hunk 20]
```diff
@@ -2,45 +2,98 @@
...
+ * @method static Builder|InternetAccess whereAutoApprovedMacSlots($value)
+ * @method static Builder|InternetAccess whereCreatedAt($value)
+ * @method static Builder|InternetAccess whereHasInternetUntil($value)
+ * @method static Builder|InternetAccess whereUpdatedAt($value)
+ * @method static Builder|InternetAccess whereUserId($value)
+ * @method static Builder|InternetAccess whereWifiPassword($value)
+ * @method static Builder|InternetAccess whereWifiUsername($value)
+ * @property-read Collection|MacAddress[] $macAddresses
+ * @property-read int|null $mac_addresses_count
+ * @mixin Eloquent
```
Comment (sszajbely) [inline app/Models/Internet/InternetAccess.php:46]: These IDE helpers would have been nice in a separate PR :D

File: app/Models/Internet/InternetAccess.php [hunk 21]
```diff
@@ -2,45 +2,98 @@
...
  */
 class InternetAccess extends Model
 {
+    protected $table = 'internet_accesses';
```
Comment (sszajbely) [inline app/Models/Internet/InternetAccess.php]: Is this our convention to state the table name in each model? Because this is not needed otherwise.

File: app/Models/Internet/MacAddress.php [hunk 22]
```diff
@@ -33,15 +55,49 @@ class MacAddress extends Model
         'user_id', 'mac_address', 'comment', 'state', 'ip',
     ];
 
+    // Default attributes
     protected $attributes = [
-        'comment' => '',
         'state' => self::REQUESTED,
     ];
```
Comment (sszajbely) [inline app/Models/Internet/MacAddress.php]: What does this do exactly? Also it would be nice to use enums instead of const strings.

File: app/Models/Internet/MacAddress.php [hunk 23]
```diff
@@ -33,15 +55,49 @@ class MacAddress extends Model
...
     {
-        return $this->belongsTo(User::class);
+        return $this->hasOneThrough(
+            User::class,
+            InternetAccess::class,
+            'user_id', // Foreign key on InternetAccess table...
+            'id', // Foreign key on Users table...
+            'user_id', // Local key on MacAddresses table...
+            'user_id' // Local key on InternetAccess table...
+        );
```
Comment (sszajbely) [inline app/Models/Internet/MacAddress.php:96]: Why? Do you want to treat the `MacAddress` as if it belonged to the `InternetAccess` even though it is not reflected in the database?

File: app/Models/Internet/MacAddress.php [hunk 24]
```diff
@@ -57,7 +113,11 @@ public function macAddress(): Attribute
         );
     }
 
-    public static function notifications()
+    /**
+     * The number of mac addresses waiting for approval.
+     * @return int
+     */
+    public static function notifications(): int
```
Comment (sszajbely) [inline app/Models/Internet/MacAddress.php]: Could the name reflect the usage? Even something like `notificationCount`.

File: app/Models/Internet/Router.php [hunk 25]
```diff
@@ -33,37 +70,30 @@ class Router extends Model
...
+     * Email the sysadmins and the residents if the router has failed for the second time.
+     * @return void
+     */
+    public function sendWarning(): void
     {
         if ($this->failed_for == self::WARNING_THRESHOLD) {
             foreach (User::admins() as $admin) {
-                Mail::to($admin)->queue(new \App\Mail\RouterWarning($admin, $this));
+                Mail::to($admin)->queue(new RouterWarning($admin, $this));
             }
```
Comment (sszajbely) [inline app/Models/Internet/Router.php:106]: Same as above

File: app/Models/Internet/Router.php [hunk 26]
```diff
@@ -33,37 +70,30 @@ class Router extends Model
...
         }
     }
 
-    public static function notifications()
+    /**
+     * Return the number of routers that have failed.
+     * @return int
+     */
+    public static function notifications(): int
```
Comment (sszajbely) [inline app/Models/Internet/Router.php]: Same as above

File: database/migrations/2023_12_25_190920_delete_internet_access_role.php [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        if (Role::where('name', 'internet-user')->first() !== null) {
+            DB::table('roles')->delete(Role::where('name', 'internet-user')->first()->id);
+        }
```
Comment (sszajbely) [inline database/migrations/2023_12_25_190920_delete_internet_access_role.php:18]:
```suggestion
        if (Role::firstWhere('name', 'internet-user') !== null) {
            DB::table('roles')->delete(Role::firstWhere('name', 'internet-user')->id);
        }
```

Or maybe even `DB::table('roles')->delete(Role::firstWhere('name', 'internet-user')?->id);` without the first if. I don't know what delete with null as id does.

**Review by sszajbely — CHANGES_REQUESTED** [review]
I didn't really look at the frontend, as I'm not really familiar with the blades. Other than my comments LGTM.

File: app/Http/Controllers/Network/AdminInternetController.php (hunk 11 above)
Comment (kdmnk) [reply app/Http/Controllers/Network/AdminInternetController.php:45]: I've tried but it seems like we need both for some reason. I haven't checked the TabulatorPaginator how we could optimise.

File: app/Http/Controllers/Network/MacAddressController.php (hunk 19 above)
Comment (kdmnk) [reply app/Http/Controllers/Network/MacAddressController.php]: If the request has mac_address or user id, this would also update that, which is a security risk.

File: app/Models/Internet/InternetAccess.php (hunk 20 above)
Comment (kdmnk) [reply app/Models/Internet/InternetAccess.php:46]: well yes, sorry

File: app/Models/Internet/InternetAccess.php (hunk 21 above)
Comment (kdmnk) [reply app/Models/Internet/InternetAccess.php]: Sometimes we need it, but true, not here.

File: app/Models/Internet/MacAddress.php (hunk 22 above)
Comment (kdmnk) [reply app/Models/Internet/MacAddress.php]: Default values are defined here instead of DB level. We don't have a convention here. I created a new issue for enums as we need to change it everywhere.

File: app/Http/Controllers/Network/MacAddressController.php (hunk 19 above)
Comment (sszajbely) [reply app/Http/Controllers/Network/MacAddressController.php]: Then the validation should reflect what you expect. $data will only contain data which was validated.

File: app/Models/Internet/MacAddress.php (hunk 23 above)
Comment (kdmnk) [reply app/Models/Internet/MacAddress.php:96]: Yes. I know its a bit strange here but I don't think this would cause problems

File: app/Models/Internet/MacAddress.php (hunk 22 above)
Comment (sszajbely) [reply app/Models/Internet/MacAddress.php]: But you can still do it here. Less to refactor later :D

File: app/Models/Internet/MacAddress.php (hunk 23 above)
Comment (sszajbely) [reply app/Models/Internet/MacAddress.php:96]: It's a bit strange and probably heavier on the queries, but I don't think it is a problem. It mainly was just a question.

File: app/Http/Controllers/Network/MacAddressController.php [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+        ]);
+
+        MacAddress::create([
+            'user_id' => user()->id,
+            'comment' => $request->input('comment'),
+            'mac_address' => $request->input('mac_address')
+        ]);
+
+        foreach (User::admins() as $admin) {
+            Mail::to($admin)->send(new MacNeedsApproval($admin->name, user()->name));
```
Comment (horcsinbalint) [inline app/Http/Controllers/Network/MacAddressController.php]: Is it possible to change in a way that new MAC addresses from admins would be automatically approved? I think we might need to sometimes test out random devices and spamming other administrators' e-mail addresses would be painful.

**Review by sszajbely — APPROVED** [review]
